src/parser.rs

```diff
@@ -13,6 +14,44 @@ use crate::{
     KdlDocument, KdlEntry, KdlError, KdlErrorKind, KdlIdentifier, KdlNode, KdlParseError, KdlValue,
 };
 
+/// The parser for the entire input.
+///
+/// All of our parsing subroutines want to hold onto some global information
+/// to generate things like spans, so instead of making them simple free
+/// functions, we wrap their bodies in closures that take in a kdl_parser.
+/// The free functions then becoming constructors that return those closures.
+/// This is basically the same idea behind nom combinators like many0 which
+/// take an input to configure the combinator and then return a function.
+pub(crate) struct KdlParser<'a> {
+    pub(crate) full_input: &'a str,
+}
+
+impl<'a> KdlParser<'a> {
+    // Creates a span for an item using two substrings of self.full_input:
+    //
+    // * before: the remainder of the input before parsing the item
+    // * after: the remainder input after parsing the item
+    //
+    // All we really care about are the addresses of the strings, the lengths don't matter
+    fn span_from_before_and_after(&self, before: &str, after: &str) -> SourceSpan {
+        let base_addr = self.full_input.as_ptr() as usize;
+        let before_addr = before.as_ptr() as usize;
+        let after_addr = after.as_ptr() as usize;
+        assert!(
+            before_addr >= base_addr,
+            "tried to get the span of a non-substring!"
+        );
+        assert!(
+            after_addr >= before_addr,
+            "subslices were in wrong order for spanning!"
+        );
+
+        let start = before_addr - base_addr;
+        let end = after_addr - base_addr;
+        SourceSpan::from(start..end)
+    }
+}
+
 pub(crate) fn parse<'a, T, P>(input: &'a str, parser: P) -> Result<T, KdlError>
 where
     P: Parser<&'a str, T, KdlParseError<&'a str>>,
```

Comment:
note: I spent a fair amount of time trying to make this take in the "higher order" parsers so that all of the KdlParsers could be built in this one spot, but the lifetimes go super wonky and rustc gets upset. I'm not sure the signature is properly expressible because of constraints around well-formedness and trait signatures. In particular associated types are assumed to have all the lifetime constraints and the Fn traits are invariant (unlike a fn ptr). Too much stuff gets unified.

As a result every caller of `parse` needs to setup the KdlParser and apply it to the top-level parsing function they want to invoke.